src/auto-complete.jsx

```diff
@@ -0,0 +1,266 @@
+              case 'string':
+                return (<MenuItem
+                          disableFocusRipple={this.props.disableFocusRipple}
+                          innerDivStyle={{overflow:'hidden'}}
+                          key={index}
+                          value={request}
+                          primaryText={request}
+                          />);
+              case 'object':
+                return React.cloneElement(request, {
+                        key: index,
+                        disableFocusRipple: this.props.disableFocusRipple,
+                      });
+              default:
+                return null;
+            }
+          })
+        }
+      </Menu>
+    ) : null;
+
+    return (
+      <div style={mergedRootStyles}
+           onKeyDown={(e)=>{
+            switch(e.keyCode){
+              case KeyCode.ESC:
+                this.setState({open:false});
+                break;
+              case KeyCode.DOWN:
+                if(this.focusOnInput && this.state.open){
+                  e.preventDefault();
+                  this.focusOnInput = false;
+                  this.setState({open:true});
+                }
+                break;
+              default:
+                break;
+            }
+          }}>
+        <div
+          style={{
+            widht:'100%',
+          }}>
+          <TextField
+            {...other}
+            ref="searchTextField"
+            value={this.state.searchText}
+            onEnterKeyDown={()=>{
+                 setTimeout(() => {
+                   this.state({open:false});
+                 }, this.props.touchTapCloseDelay);
+                this.props.onNewRequest(this.state.searchText);
+            }}
+            onChange={(e)=>{
+                let searchText = e.target.value;
+                this.updateRequests(searchText);
+            }}
+            onBlur={()=>{
+              if(this.focusOnInput && this.state.open)
+                this.refs.searchTextField.focus();
+            }}
+            onFocus={()=>{
+              if(!this.state.open && (
+                  this.props.updateWhenFocused || this.searchText !== '')){
+                this.updateRequests(this.state.searchText);
+              }
+              this.focusOnInput = true;
+            }}
+
+            {...textFieldProps} />
+        </div>
+          <ReactTransitionGroup>{menu}</ReactTransitionGroup>
+      </div>
+    );
+  },
+
+  setValue(textValue) {
+    this.setState({
+      searchText: textValue,
+    });
+  },
+
+  getValue() {
+    return this.state.searchText;
+  },
+
+  updateRequests(searchText){
+    let requestsListPromise = new Promise((resolve, reject) => {
+      let list = this.props.onUpdateRequests(searchText);
+      if(list){
+        resolve(list);
+      }
+      else{
+        reject();
+      }
+    });
+
+    requestsListPromise.then(list => {
+      this.focusOnInput = true;
+      this.setState({
+        searchText:searchText,
```

Comment:
`searchText` state should probably be set independently of the `Promise` result. I found that if the server is slow to respond, then the UI does not reflect what the user typed in until _after_ the server has responded in someway. Weirder still, when the server requests return in a different order, then the rendered value can regress to a previous value.

I would recommend something like: 
```
    this.setState({
      searchText: searchText,
    });
    requestsListPromise.then(list => {
      // ...
    })
```

This lets the UI redraw to reflect the user's input immediately.